### PR TITLE
global: metadata extensions

### LINKED ISSUE #8: UI override components to support extra metadata
https://github.com/CERNDocumentServer/cds-books/issues/82

If the overlay provides extra metadata for the records, we have to be able to override default components so we can display/edit the extra metadata of the record.

- [x] Search page of the record
- [x] Details page of the record
- [x] Form of the record

### PR DESCRIPTION
**Pull request by topless**
- searchConfig.js sort option for acqVendors and IllLibaries as keyword
- fix props types for error
- fix for document details relations
- added conditional Tab pane for document details 
- added override document metadata extensions component
- no support for mobile
- eslint fixes
- implemented backoffice search document extensions overridable
- Document & Series Identifiers fixed and simplified
- moved overridable components to their own module
- document form support for overridable extensions fields
- errors as default null
- removed extensions from doc search
- configurable label for extensions
- fixed date field
- replaced components with type of value
- moved default implementations inside ILS
- show/hide fields by config
- added mobile view
- fix mobile view to match desktop view content
- restored custom components
- throw error as default instead of StringField
- restored custom components
- throw error as default instead of StringField
- added series components
- fixes for props on backoffice Document & Series details
- tests
- Overridable components with explicit props
- removed the unused class `document-details-search-bar` 
- renamed css class `document-details-search-container` to `literature-search-container` since its used in series too.
- requires #57
- closes #8

Co-authored-by: Nicola <ntarocco@gmail.com>

### CODE REVIEW

File: src/lib/pages/backoffice/Document/DocumentForms/DocumentEditor/DocumentForm/DocumentForm.js [hunk 1]
```diff
@@ -13,24 +13,25 @@ import { BackOfficeRoutes } from '@routes/urls';
...
   AlternativeAbstracts,
   AlternativeIdentifiers,
   AlternativeTitles,
   AuthorsField,
+  ConferenceInfoField,
   Copyrights,
   Identifiers,
+  Imprint,
+  InternalNotes,
+  Keywords,
```
Comment (ntarocco) [inline src/lib/pages/backoffice/Document/DocumentForms/DocumentEditor/DocumentForm/DocumentForm.js:28]: Watch out for this `index.js` that will not trigger a circular dependency

File: src/overrides/DocumentFormExtensions.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import PropTypes from 'prop-types';
```
Comment (ntarocco) [inline src/overrides/DocumentFormExtensions.js:1]: Is `src/overrides` a new folder? What is the purpose of it?

Comment (topless) [reply src/overrides/DocumentFormExtensions.js:1]: It is **temporary**. The `index.js` where I was putting all the override implementations started to grow, so I put them all in the ovverides folder. The purpose of it is to disappear and move its components to overlay.

File: src/lib/pages/backoffice/Document/DocumentForms/DocumentEditor/DocumentForm/DocumentForm.js (hunk 1 above)
Comment (topless) [reply src/lib/pages/backoffice/Document/DocumentForms/DocumentEditor/DocumentForm/DocumentForm.js:28]:
```
Processed 998 files (6.5s) (281 warnings)

✔ No circular dependency found!
```

**Comment by ntarocco** [comment]
> Thanks for the code changes! I have a general comment because maybe I was not understanding your idea for the implementation.
> Since for the backend you "just" configure custom fields, I was expecting the same for the UI. Instead, you have to write your own React components and logic.
> It would be nice to discuss with the team what, as a user/developer, we expect.

As discussed, it is just a matter to move the example overridden components as default implementation.
Good job! Looks good! 👏

File: src/extensionsConfig.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
+import _keys from 'lodash/keys';
+
+const fields = {
```
Comment (ntarocco) [inline src/extensionsConfig.js:3]: is this an example? If yes, I would keep only something random, e.g. `example_extra_field` -> string, and keep all this for CDS.
I would not put CERN/CDS stuff in InvenioILS.

WDYT?

File: src/index.js [hunk 4]
```diff
@@ -2,34 +2,10 @@ import React from 'react';
...
-      size="tiny"
-    />
-  );
-};
-
-const overriddenCmps = {
-  // 'Home.layout': CustomHome,
-  // 'LiteratureCover.layout': CustomCover,
-};
+const overriddenCmps = {};
```
Comment (ntarocco) [inline src/index.js:8]: I believe you have removed this because of the warnings in console, but it is handy to test if overridable still works and as example.
What about if you comment it out instead?

File: src/lib/config/index.js [hunk 5]
```diff
@@ -1,3 +1,6 @@
 export { getSearchConfig } from './searchConfig';
 export { invenioConfig } from './invenioConfig';
 export { uiConfig } from './uiConfig';
+
+// NOTE: Support for external configuration, merge with defaults
```
Comment (ntarocco) [inline src/lib/config/index.js:5]: not sure I understand the comment, in any case you will have to refactor it when providing overridable config...

File: src/lib/pages/backoffice/Document/DocumentForms/DocumentEditor/DocumentForm/components/MetadataExtensions.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+const getFormComponent = fieldType => {
+  switch (fieldType) {
+    case 'string':
+      return StringField;
+    case 'boolean':
+      return BooleanField;
+    case 'date':
+      return DateInputField;
+    default:
+      return StringField;
```
Comment (ntarocco) [inline src/lib/pages/backoffice/Document/DocumentForms/DocumentEditor/DocumentForm/components/MetadataExtensions.js:21]: I would throw as default maybe? WDYT?

File: src/lib/pages/backoffice/Document/DocumentForms/DocumentEditor/DocumentForm/components/MetadataExtensions.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  }
+};
+
+export const MetadataExtensions = ({ extensions }) => {
+  extensions = _defaultsDeep(extensions, extensionsConfig.defaults());
+
+  let fields = [];
+  _forOwn(extensions, (value, key) => {
+    const componentType = _get(extensionsConfig.fields, `${key}.type`);
```
Comment (ntarocco) [inline src/lib/pages/backoffice/Document/DocumentForms/DocumentEditor/DocumentForm/components/MetadataExtensions.js:30]:
```suggestion
    const componentType = extensionsConfig.fields[key].type
```

File: src/lib/pages/backoffice/Document/DocumentForms/DocumentEditor/DocumentForm/components/MetadataExtensions.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    const Component = getFormComponent(componentType);
+
+    fields.push(
+      <Component
+        fieldPath={`extensions.${key}`}
+        key={key}
+        label={key}
+        optimized
+        required={_get(extensionsConfig.fields, `${key}.isRequired`)}
```
Comment (ntarocco) [inline src/lib/pages/backoffice/Document/DocumentForms/DocumentEditor/DocumentForm/components/MetadataExtensions.js:39]: very minor, just for coherence:
```suggestion
        required={_get(extensionsConfig.fields, `${key}.isRequired`, false)}
```
or
```suggestion
        required={'isRequired' in extensionsConfig.fields[key] ? extensionsConfig.fields[key].isRequired : false}
```

File: src/lib/pages/frontsite/Documents/DocumentDetails/DocumentItems/DocumentItems.js [hunk 9]
```diff
@@ -92,13 +93,8 @@ class DocumentItems extends Component {
...
     const { activeLibrary } = this.state;
-    const {
-      documentDetails: {
-        metadata: {
-          items: { on_shelf: onShelf },
-        },
-      },
-    } = this.props;
+    const { documentDetails } = this.props;
+    const onShelf = _get(documentDetails, 'metadata.items.on_shelf', {});
```
Comment (ntarocco) [inline src/lib/pages/frontsite/Documents/DocumentDetails/DocumentItems/DocumentItems.js:97]: Maybe I am mistaken, but didn't we said remote the `_get` here?

File: src/lib/pages/frontsite/Documents/DocumentDetails/DocumentMetadata/DocumentMetadataAccordion.js [hunk 10]
```diff
@@ -37,6 +44,18 @@ class DocumentMetadataAccordion extends Component {
           <DocumentInfo metadata={metadata} />
         </Accordion.Content>
 
+        <Accordion.Title
```
Comment (ntarocco) [inline src/lib/pages/frontsite/Documents/DocumentDetails/DocumentMetadata/DocumentMetadataAccordion.js:47]: 👍

File: src/extensionsConfig.js [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    default: '',
+  },
+  'accelerator_experiments:study': {
+    type: 'string',
+    default: '',
+  },
+  'standard_CERN_status:CERN_applicability': {
+    type: 'string',
+    default: '',
+    isVisible: false,
```
Comment (ntarocco) [inline src/extensionsConfig.js:35]: why the  `isVisible`?

File: src/lib/pages/frontsite/Documents/DocumentDetails/DocumentMetadata/DocumentMetadataExtensions.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+export const DocumentMetadataExtensions = ({ extensions }) => {
+  extensions = _pickBy(extensions, (value, key) => {
+    return _get(extensionsConfig.fields[key], 'isVisible', true);
+  });
+
+  let result = {};
+
+  _keys(extensions).map(key => {
+    const [objName, objProp] = key.split(':');
```
Comment (ntarocco) [inline src/lib/pages/frontsite/Documents/DocumentDetails/DocumentMetadata/DocumentMetadataExtensions.js:18]: I am not 100% sure I understand, but is this splitting done to avoid the `:`? Is the field name that we send to the backend `accelerator_experiments:accelerator` or only `accelerator`? Or why this split?
Would it be possible to make this manipulation a little bit more readable/understandable?


I would avoid any manipulation to display a label, as I was commenting this is what is done in `invenio-search-js/ui` and it really does not work. What we should do instead is to add a `label` field in the config so that the user can customize what name to display.

File: src/extensionsConfig.js (hunk 3 above)
Comment (topless) [reply src/extensionsConfig.js:3]: This is the actual CDS config, will be moved in CDS

File: src/extensionsConfig.js (hunk 11 above)
Comment (topless) [reply src/extensionsConfig.js:35]: `isVisible` controls if the field will be displayed in the frontsite to the Patron. If we do not configure the field, then we cannot show it in the backoffice form. I am more than open to ideas about controlling that behaviour.

File: src/lib/pages/frontsite/Documents/DocumentDetails/DocumentItems/DocumentItems.js (hunk 9 above)
Comment (topless) [reply src/lib/pages/frontsite/Documents/DocumentDetails/DocumentItems/DocumentItems.js:97]: You are right 👍

File: src/lib/pages/frontsite/Documents/DocumentDetails/DocumentMetadata/DocumentMetadataExtensions.js (hunk 12 above)
Comment (topless) [reply src/lib/pages/frontsite/Documents/DocumentDetails/DocumentMetadata/DocumentMetadataExtensions.js:18]: The convention used in the backend for extensions is `{'<namespace>:<field>': <value> }`, so our data look like

```js
extensions: {
  accelerator_experiments:accelerator: 'AccelerateMe',
  accelerator_experiments:project: 'CDS',
}
```
and so on. 

So here I am transforming the object to the following form to avoid rendering repetitive information.  
```
extensions: {
  accelerator_experiments: {
    accelerator: 'AccelerateMe',
    project: 'CDS',
  }
}
```
I think that moving it in the `api/document/serializer` and also deserialize when we submit DocumentForm is cleaner and will hold the conversion in one place only.

Dropping some of my thoughts, then this will not match with our ui field configuration structure, which if it gets changed will not match with the backend field configuration structure.

@ntarocco wdyt?

File: src/lib/modules/Series/SeriesTitle.js [hunk 13]
```diff
@@ -10,6 +10,11 @@ export const SeriesTitle = ({ title, modeOfIssuance }) => (
 );
 
 SeriesTitle.propTypes = {
-  title: PropTypes.string.isRequired,
-  modeOfIssuance: PropTypes.string.isRequired,
+  title: PropTypes.string,
```
Comment (ntarocco) [inline src/lib/modules/Series/SeriesTitle.js:13]: actually these are required in the `schema`, they must be there... I would revert this change maybe

File: src/lib/pages/backoffice/Document/DocumentForms/DocumentEditor/DocumentForm/components/MetadataExtensions.js [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+  let components = [];
+  _forOwn(allExtensions, (value, key) => {
+    const componentType = extensionsConfig.document.fields[key].type;
+    const Component = getFormComponent(componentType);
+
+    components.push(
+      <Component
+        fieldPath={`extensions.${key}`}
+        key={key}
+        label={key}
```
Comment (ntarocco) [inline src/lib/pages/backoffice/Document/DocumentForms/DocumentEditor/DocumentForm/components/MetadataExtensions.js:43]:
```suggestion
        label={_get(extensionsConfig.document.fields, `${key}.label`, key)}
```

File: src/lib/pages/backoffice/Series/SeriesDetails/SeriesDetails.js [hunk 15]
```diff
@@ -94,7 +93,10 @@ export default class SeriesDetails extends Component {
       content: (
         <Accordion.Content>
           <div id="series-monographs">
-            <SeriesMultipartMonographs />
+            <SeriesMultipartMonographs
+              seriesPid={_get(data, 'pid')}
```
Comment (ntarocco) [inline src/lib/pages/backoffice/Series/SeriesDetails/SeriesDetails.js:97]: we probably have changed the same, in any case `pid` and `mode_of_issuance` are required by the JSON schema, so you don't need the `_get`

File: src/lib/pages/backoffice/Series/SeriesForms/SeriesEditor/components/SeriesForm/SeriesMetadataExtensions.js [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+      return DateInputField;
+    default:
+      throw new Error(
+        `${fieldType} did not match any of the supported types: string, date or boolean`
+      );
+  }
+};
+
+export const SeriesMetadataExtensions = ({ extensions }) => {
+  const { label, fields } = extensionsConfig.series;
```
Comment (ntarocco) [inline src/lib/pages/backoffice/Series/SeriesForms/SeriesEditor/components/SeriesForm/SeriesMetadataExtensions.js:29]: Did it make sense to create a re-usable form component for both Document and Series instead?

File: src/lib/pages/frontsite/Documents/DocumentDetails/DocumentDetails.js [hunk 17]
```diff
@@ -156,7 +160,7 @@ class DocumentDetails extends Component {
     const { searchQuery } = this.state;
     return (
       <>
-        <Overridable id="DocumentDetails.top">
+        <Overridable id="DocumentDetails.top" searchQuery={searchQuery}>
```
Comment (ntarocco) [inline src/lib/pages/frontsite/Documents/DocumentDetails/DocumentDetails.js:163]: This is for sure needed, but not enough: there are also these 2 functions `this.onSearchInputChange` and `this.onSearchClick` that are not callable when overriding.
The interface when overriding should be:
```
(currentQueryString, onInputChange, executeSearch) => {}
```
which probably means:
```
<Overridable id="DocumentDetails.top" currentQueryString={searchQuery} onInputChange={this.onSearchInputChange} executeSearch={this.onSearchClick}>
...
              <SearchBar
                currentQueryString={currentQueryString}
                onInputChange={onInputChange}
                executeSearch={executeSearch}
                placeholder="Search for books..."
                className="document-details-search-bar"
              />
...
```
to be tested...

It would be nice also to change `"Search for books..."` to `"Search for literatures..."`

File: src/lib/pages/frontsite/Series/SeriesDetails/SeriesPanel/SeriesSequences.js [hunk 18]
```diff
@@ -63,7 +63,11 @@ class SeriesSequences extends Component {
 }
 
 SeriesSequences.propTypes = {
-  relations: PropTypes.object.isRequired,
+  relations: PropTypes.object,
```
Comment (ntarocco) [inline src/lib/pages/frontsite/Series/SeriesDetails/SeriesPanel/SeriesSequences.js:66]: Are you sure about this change? In general, IMHO, we should have required as much as possible, and limit the use of `_get` to what is really optional.
If this component should not render when there are no relations, it should be handled when calling it...

Comment (topless) [reply src/lib/pages/frontsite/Series/SeriesDetails/SeriesPanel/SeriesSequences.js:66]: When I am accessing a `Serial` I get Warning: Failed prop type: The prop `relations` is marked as required in `SeriesSequences`, but its value is `undefined`. That is why I changed it, I can revert it back and handle all these props issues together in an issue.

File: src/lib/pages/frontsite/Documents/DocumentDetails/DocumentDetails.js (hunk 17 above)
Comment (topless) [reply src/lib/pages/frontsite/Documents/DocumentDetails/DocumentDetails.js]: I have tested this (with a custom overridable component)

- I have updated the Overridable components with explicit props
- I have updated the placeholders
- removed the unused class `document-details-search-bar` 
- renamed the style class to `literature` instead of documents since its used in series too.

File: src/lib/pages/backoffice/Series/SeriesForms/SeriesEditor/components/SeriesForm/SeriesMetadataExtensions.js (hunk 16 above)
Comment (topless) [reply src/lib/pages/backoffice/Series/SeriesForms/SeriesEditor/components/SeriesForm/SeriesMetadataExtensions.js:29]: yes it makes, created an issue #69

File: src/lib/modules/Series/SeriesTitle.js (hunk 13 above)
Comment (topless) [reply src/lib/modules/Series/SeriesTitle.js:13]: I was carried away from series minimal which actually derives from another error, #70
